base/stream.jl

```diff
@@ -1215,20 +1215,118 @@ i.e. data to be read from [`stdin`](@ref) may be written to `wr`.
 """
 redirect_stdin
 
-for (F,S) in ((:redirect_stdin, :stdin), (:redirect_stdout, :stdout), (:redirect_stderr, :stderr))
-    @eval function $F(f::Function, stream)
-        STDOLD = $S
-        local ret
-        $F(stream)
-        try
-            ret = f()
-        finally
-            $F(STDOLD)
-        end
-        ret
+
+"""
+    redirect(;stdin=stdin, stderr=stderr, stdout=stdout)
+
+Redirect a subset of the streams `stdin`, `stderr`, `stdout`.
+"""
+function redirect(;stdin=nothing, stderr=nothing, stdout=nothing)
+    stdin  === nothing || redirect_stdin(stdin)
+    stderr === nothing || redirect_stderr(stderr)
+    stdout === nothing || redirect_stdout(stdout)
```

Comment:
It seems to me the path and error checking should be moved here, so that you get them with or without a function argument?

Reply:
Note that this function does not accept paths. The reason is, that while this function could open the corresponding file, it cannot close it. So there is no error hadling that can be done here.